README.md
# Overview

Python module that aids in making and publishing patterns for sewing projects

Inspired by the [freesewing](http://freesewing.org) project.

This module is still in the planning phases. Module name is not final

This project was generated with [cookiecutter](https://github.com/audreyr/cookiecutter) using [jacebrowning/template-python](https://github.com/jacebrowning/template-python).

[![Unix Build Status](https://img.shields.io/travis/destos/patterner/master.svg?label=unix)](https://travis-ci.org/destos/patterner)
[![Windows Build Status](https://img.shields.io/appveyor/ci/destos/patterner/master.svg?label=window)](https://ci.appveyor.com/project/destos/patterner)
[![Coverage Status](https://img.shields.io/coveralls/destos/patterner/master.svg)](https://coveralls.io/r/destos/patterner)
[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/destos/patterner.svg)](https://scrutinizer-ci.com/g/destos/patterner/?branch=master)
[![PyPI Version](https://img.shields.io/pypi/v/Patterner.svg)](https://pypi.org/project/Patterner)
[![PyPI License](https://img.shields.io/pypi/l/Patterner.svg)](https://pypi.org/project/Patterner)

# Setup

## Requirements

* Python 3.7+

## Installation

Install this library directly into an activated virtual environment:

```bash
$ pip install patterner
```

or add it to your [Poetry](https://poetry.eustace.io/) project:

```bash
$ poetry add patterner
```

# Usage

After installation, the package can be imported:

```bash
$ python
>>> import patterner
>>> patterner.__version__
```


CHANGELOG.md
# 0.1.0 (2019-03-01)

 - Initial release

Let's get this party started.


tests/conftest.py
"""Integration tests configuration file."""

# pylint: disable=unused-import

from patterner.tests.conftest import pytest_configure


tests/test_cli.py
"""Sample integration test module using pytest-describe and expecter."""
# pylint: disable=redefined-outer-name,unused-variable,expression-not-assigned

import pytest
from click.testing import CliRunner
from expecter import expect

from patterner.cli import main


@pytest.fixture
def runner():
    return CliRunner()


def describe_cli():
    def describe_conversion():
        def when_integer(runner):
            result = runner.invoke(main, ['42'])

            expect(result.exit_code) == 0
            expect(result.output) == "12.80165\n"

        def when_invalid(runner):
            result = runner.invoke(main, ['foobar'])

            expect(result.exit_code) == 0
            expect(result.output) == ""


patterner/tests/test_models.py
import unittest

from patterner import models


class TestModels(unittest.TestCase):
    def test_dataclass_params(self):
        params = models.BodyModel.__dataclass_params__
        self.assertTrue(params.frozen)
        self.assertTrue(params.init)
        self.assertTrue(params.repr)
        self.assertTrue(params.eq)
        self.assertFalse(params.order)
        self.assertFalse(params.unsafe_hash)

    def test_body_model_attibutes_defaults(self):
        model = models.BodyModel()
        self.assertEqual(model.biceps_circumference, 0)
        self.assertEqual(model.center_back_neck_to_waist, 0)
        self.assertEqual(model.chest_circumference, 0)
        self.assertEqual(model.head_circumference, 0)
        self.assertEqual(model.hips_circumference, 0)
        self.assertEqual(model.hips_to_upper_leg, 0)
        self.assertEqual(model.natural_waist, 0)
        self.assertEqual(model.natural_waist_to_hip, 0)
        self.assertEqual(model.neck_circumference, 0)
        self.assertEqual(model.shoulder_slope, 0)
        self.assertEqual(model.shoulder_to_elbow, 0)
        self.assertEqual(model.shoulder_to_shoulder, 0)
        self.assertEqual(model.shoulder_to_wrist, 0)
        self.assertEqual(model.upper_leg_circumference, 0)
        self.assertEqual(model.wrist_circumference, 0)

    def test_man_size_34(self):
        model = models.man_size_34
        self.assertEqual(model.biceps_circumference, 335)
        self.assertEqual(model.center_back_neck_to_waist, 489)
        self.assertEqual(model.chest_circumference, 849)
        self.assertEqual(model.head_circumference, 570)
        self.assertEqual(model.hips_circumference, 722)
        self.assertEqual(model.hips_to_upper_leg, 183)
        self.assertEqual(model.natural_waist_to_hip, 100)
        self.assertEqual(model.neck_circumference, 366)
        self.assertEqual(model.shoulder_slope, 43)
        self.assertEqual(model.shoulder_to_shoulder, 419)
        self.assertEqual(model.shoulder_to_wrist, 670)
        self.assertEqual(model.upper_leg_circumference, 565)
        self.assertEqual(model.wrist_circumference, 175)

    def test_man_size_36(self):
        model = models.man_size_36
        self.assertEqual(model.biceps_circumference, 290)
        self.assertEqual(model.center_back_neck_to_waist, 492)
        self.assertEqual(model.chest_circumference, 907)
        self.assertEqual(model.head_circumference, 575)
        self.assertEqual(model.hips_circumference, 780)
        self.assertEqual(model.hips_to_upper_leg, 193)
        self.assertEqual(model.natural_waist_to_hip, 105)
        self.assertEqual(model.neck_circumference, 378)
        self.assertEqual(model.shoulder_slope, 46)
        self.assertEqual(model.shoulder_to_shoulder, 431)
        self.assertEqual(model.shoulder_to_wrist, 675)
        self.assertEqual(model.upper_leg_circumference, 582)
        self.assertEqual(model.wrist_circumference, 180)

    def test_man_size_38(self):
        model = models.man_size_38
        self.assertEqual(model.biceps_circumference, 305)
        self.assertEqual(model.center_back_neck_to_waist, 495)
        self.assertEqual(model.chest_circumference, 965)
        self.assertEqual(model.head_circumference, 580)
        self.assertEqual(model.hips_circumference, 838)
        self.assertEqual(model.hips_to_upper_leg, 202)
        self.assertEqual(model.natural_waist_to_hip, 110)
        self.assertEqual(model.neck_circumference, 391)
        self.assertEqual(model.shoulder_slope, 49)
        self.assertEqual(model.shoulder_to_shoulder, 444)
        self.assertEqual(model.shoulder_to_wrist, 680)
        self.assertEqual(model.upper_leg_circumference, 598)
        self.assertEqual(model.wrist_circumference, 185)

    def test_man_size_40(self):
        model = models.man_size_40
        self.assertEqual(model.biceps_circumference, 320)
        self.assertEqual(model.center_back_neck_to_waist, 498)
        self.assertEqual(model.chest_circumference, 1023)
        self.assertEqual(model.head_circumference, 585)
        self.assertEqual(model.hips_circumference, 896)
        self.assertEqual(model.hips_to_upper_leg, 211)
        self.assertEqual(model.natural_waist_to_hip, 115)
        self.assertEqual(model.neck_circumference, 404)
        self.assertEqual(model.shoulder_slope, 52)
        self.assertEqual(model.shoulder_to_shoulder, 457)
        self.assertEqual(model.shoulder_to_wrist, 685)
        self.assertEqual(model.upper_leg_circumference, 614)
        self.assertEqual(model.wrist_circumference, 190)

    def test_man_size_42(self):
        model = models.man_size_42
        self.assertEqual(model.biceps_circumference, 335)
        self.assertEqual(model.center_back_neck_to_waist, 501)
        self.assertEqual(model.chest_circumference, 1081)
        self.assertEqual(model.head_circumference, 590)
        self.assertEqual(model.hips_circumference, 995)
        self.assertEqual(model.hips_to_upper_leg, 220)
        self.assertEqual(model.natural_waist, 925)
        self.assertEqual(model.natural_waist_to_hip, 120)
        self.assertEqual(model.neck_circumference, 416)
        self.assertEqual(model.shoulder_slope, 55)
        self.assertEqual(model.shoulder_to_elbow, 415)
        self.assertEqual(model.shoulder_to_shoulder, 470)
        self.assertEqual(model.shoulder_to_wrist, 690)
        self.assertEqual(model.upper_leg_circumference, 630)
        self.assertEqual(model.wrist_circumference, 195)

    def test_man_size_44(self):
        model = models.man_size_44
        self.assertEqual(model.biceps_circumference, 350)
        self.assertEqual(model.center_back_neck_to_waist, 505)
        self.assertEqual(model.chest_circumference, 1139)
        self.assertEqual(model.head_circumference, 595)
        self.assertEqual(model.hips_circumference, 1012)
        self.assertEqual(model.hips_to_upper_leg, 229)
        self.assertEqual(model.natural_waist_to_hip, 125)
        self.assertEqual(model.neck_circumference, 429)
        self.assertEqual(model.shoulder_slope, 58)
        self.assertEqual(model.shoulder_to_shoulder, 483)
        self.assertEqual(model.shoulder_to_wrist, 700)
        self.assertEqual(model.upper_leg_circumference, 646)
        self.assertEqual(model.wrist_circumference, 200)


scent.py
"""Configuration file for sniffer."""

import time
import subprocess

from sniffer.api import select_runnable, file_validator, runnable
try:
    from pync import Notifier
except ImportError:
    notify = None
else:
    notify = Notifier.notify


watch_paths = ["patterner", "tests"]


class Options:
    group = int(time.time())  # unique per run
    show_coverage = False
    rerun_args = None

    targets = [
        (('make', 'test-unit', 'DISABLE_COVERAGE=true'), "Unit Tests", True),
        (('make', 'test-all'), "Integration Tests", False),
        (('make', 'check'), "Static Analysis", True),
        (('make', 'docs'), None, True),
    ]


@select_runnable('run_targets')
@file_validator
def python_files(filename):
    return filename.endswith('.py')


@select_runnable('run_targets')
@file_validator
def html_files(filename):
    return filename.split('.')[-1] in ['html', 'css', 'js']


@runnable
def run_targets(*args):
    """Run targets for Python."""
    Options.show_coverage = 'coverage' in args

    count = 0
    for count, (command, title, retry) in enumerate(Options.targets, start=1):

        success = call(command, title, retry)
        if not success:
            message = "✅ " * (count - 1) + "❌"
            show_notification(message, title)

            return False

    message = "✅ " * count
    title = "All Targets"
    show_notification(message, title)
    show_coverage()

    return True


def call(command, title, retry):
    """Run a command-line program and display the result."""
    if Options.rerun_args:
        command, title, retry = Options.rerun_args
        Options.rerun_args = None
        success = call(command, title, retry)
        if not success:
            return False

    print("")
    print("$ %s" % ' '.join(command))
    failure = subprocess.call(command)

    if failure and retry:
        Options.rerun_args = command, title, retry

    return not failure


def show_notification(message, title):
    """Show a user notification."""
    if notify and title:
        notify(message, title=title, group=Options.group)


def show_coverage():
    """Launch the coverage report."""
    if Options.show_coverage:
        subprocess.call(['make', 'read-coverage'])

    Options.show_coverage = False


patterner/patterns.py
from itertools import repeat
from operator import add, sub
from typing import NamedTuple


Infinity = float('inf')


class Point(NamedTuple):
    x: int
    y: int

    def _operation(self, other, op):
        def do_op(the_other):
            # Do you know what this is doing?
            return Point(*tuple(map(op, self, the_other)))

        if isinstance(other, Point):
            return do_op(other)
        elif isinstance(other, int):
            other = tuple(repeat(other, 2))
            return do_op(other)
        else:
            raise TypeError

    def __add__(self, other):
        return self._operation(other, add)

    def __sub__(self, other):
        return self._operation(other, sub)

    def __str__(self):
        return ""

    @property
    def real(self):
        """"Proxy due to silly use of complex type by svg.path"""
        return self.x

    @property
    def imag(self):
        """"Proxy due to silly use of complex type by svg.path"""
        return self.y



patterner/models.py
"""
models of different common body measurements and sizes
"""
from dataclasses import dataclass


@dataclass(frozen=True)
class BodyModel:
    biceps_circumference: int = 0
    center_back_neck_to_waist: int = 0
    chest_circumference: int = 0
    head_circumference: int = 0
    hips_circumference: int = 0
    hips_to_upper_leg: int = 0
    natural_waist: int = 0
    natural_waist_to_hip: int = 0
    neck_circumference: int = 0
    shoulder_slope: int = 0
    shoulder_to_elbow: int = 0
    shoulder_to_shoulder: int = 0
    shoulder_to_wrist: int = 0
    upper_leg_circumference: int = 0
    wrist_circumference: int = 0


class FemaleBodyModel(BodyModel):
    """Includes extra parameters for suspiciously shapely individuals"""

    # TODO: actually include extra fields


man_size_34 = BodyModel(
    biceps_circumference=335,
    center_back_neck_to_waist=489,
    chest_circumference=849,
    head_circumference=570,
    hips_circumference=722,
    hips_to_upper_leg=183,
    natural_waist_to_hip=100,
    neck_circumference=366,
    shoulder_slope=43,
    shoulder_to_shoulder=419,
    shoulder_to_wrist=670,
    upper_leg_circumference=565,
    wrist_circumference=175,
)


man_size_36 = BodyModel(
    biceps_circumference=290,
    center_back_neck_to_waist=492,
    chest_circumference=907,
    head_circumference=575,
    hips_circumference=780,
    hips_to_upper_leg=193,
    natural_waist_to_hip=105,
    neck_circumference=378,
    shoulder_slope=46,
    shoulder_to_shoulder=431,
    shoulder_to_wrist=675,
    upper_leg_circumference=582,
    wrist_circumference=180,
)


man_size_38 = BodyModel(
    biceps_circumference=305,
    center_back_neck_to_waist=495,
    chest_circumference=965,
    head_circumference=580,
    hips_circumference=838,
    hips_to_upper_leg=202,
    natural_waist_to_hip=110,
    neck_circumference=391,
    shoulder_slope=49,
    shoulder_to_shoulder=444,
    shoulder_to_wrist=680,
    upper_leg_circumference=598,
    wrist_circumference=185,
)


man_size_40 = BodyModel(
    biceps_circumference=320,
    center_back_neck_to_waist=498,
    chest_circumference=1023,
    head_circumference=585,
    hips_circumference=896,
    hips_to_upper_leg=211,
    natural_waist_to_hip=115,
    neck_circumference=404,
    shoulder_slope=52,
    shoulder_to_shoulder=457,
    shoulder_to_wrist=685,
    upper_leg_circumference=614,
    wrist_circumference=190,
)


man_size_42 = BodyModel(
    biceps_circumference=335,
    center_back_neck_to_waist=501,
    chest_circumference=1081,
    head_circumference=590,
    hips_circumference=995,
    hips_to_upper_leg=220,
    natural_waist=925,
    natural_waist_to_hip=120,
    neck_circumference=416,
    shoulder_slope=55,
    shoulder_to_elbow=415,
    shoulder_to_shoulder=470,
    shoulder_to_wrist=690,
    upper_leg_circumference=630,
    wrist_circumference=195,
)


man_size_44 = BodyModel(
    biceps_circumference=350,
    center_back_neck_to_waist=505,
    chest_circumference=1139,
    head_circumference=595,
    hips_circumference=1012,
    hips_to_upper_leg=229,
    natural_waist_to_hip=125,
    neck_circumference=429,
    shoulder_slope=58,
    shoulder_to_shoulder=483,
    shoulder_to_wrist=700,
    upper_leg_circumference=646,
    wrist_circumference=200,
)
